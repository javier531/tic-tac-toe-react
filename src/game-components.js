import React from 'react';
import styled from "styled-components";

const StyledSquare = styled.button`
    color: ${props => props.value === "X" ? "green" : "blue"}; 
    background: #fff;
    border: 1px solid #999;
    float: left;
    font-size: 48px;
    font-weight: bold;
    line-height: 48px;
    height: 64px;
    width: 64px;
    margin-right: -1px;
    margin-top: -1px;
    padding: 0;
    text-align: center;

    &:focus {
        outline: none;
    }
`;

export function Square(props) {
    return (
        <StyledSquare className="square" onClick={props.onClick} value={props.value}>
            {props.value}
        </StyledSquare>
    );
  }
  


export function Board(props) {

    function renderSquare(i) {
        return <Square value={props.squares[i]}
            onClick={() => props.onClick(i)} />;
    }


    let board = [];
    for (let i=0; i < 3; i++) {
        let items = [];
        for (let j=i*3; j < (i+1)*3; j++) {
            items.push(renderSquare(j));
        }
        board.push(<div className="board-row">{items}</div>);
    }
    
  
    return (
          <div>
            {board}
          </div>
    );
  }