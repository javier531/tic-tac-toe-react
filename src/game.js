import React from 'react';
import {Board} from './game-components';
import {ToggleButtonOnOff} from './ui-components';
import {calculateWinner} from './game-functions';
import GlobalStyle from './styles';

export class Game extends React.Component {
    
    constructor(props) {
        super(props);
        this.state = {
          history: [{
            squares: Array(9).fill(null),
            index: -1,
          }],
          stepNumber: 0,
          xIsNext: true,
          historyOrder: 'ASC',
        };
    }

    handleClick(i) {
        const history = this.state.history.slice(0, this.state.stepNumber + 1);
        const current = history[history.length - 1];
        const squares = current.squares.slice();

        if (calculateWinner(squares) || squares[i]) {
          return;
        }

        squares[i] = this.state.xIsNext ? 'X' : 'O';
        this.setState({
          history: history.concat([{
            squares: squares,
            index: i
          }]),
          stepNumber: history.length,
          xIsNext: !this.state.xIsNext,
        });
    }

    jumpTo(step) {
        this.setState({
          stepNumber: step,
          xIsNext: (step % 2) === 0,
        });
    }

    changeHistoryOrder() {
        this.setState({
          historyOrder: this.state.historyOrder === 'ASC' ? 'DESC' : 'ASC',
        });
    }

    render() {
        const order = this.state.historyOrder;
        const history = (order === 'ASC') ? this.state.history : [...this.state.history].reverse();
        const current = this.state.history[this.state.stepNumber];
        const winner = calculateWinner(current.squares);

        const moves = history.map((step, move) => {
            const index = step.index;
            const row = Math.floor(index/3)+1;
            const col = index +1 - ((row-1)*3);
            const listIndex = (order === 'ASC') ? move : history.length - move -1;
            const description = index > -1 ?
              'Ir a #' + listIndex + ' (' + row + ';' + col + ')':
              'Reiniciar';
            const style = this.state.stepNumber === listIndex ? 'last-move' : '';
            return (
              <li key={index}>
                <button onClick={() => this.jumpTo(listIndex)}
                    className={style}>{description}</button>
              </li>
            );
          });

        let status;
        if (winner) {
            status = 'Ganador: ' + winner;
        } else if(this.state.stepNumber === 9) {
            status = 'Empate!';
        } else {
            status = 'PrÃ³ximo: ' + (this.state.xIsNext ? 'X' : 'O');
        }

    
        return (
            
            <div className="game">
                <GlobalStyle />
                <div className="game-board">
                <Board
                    squares={current.squares}
                    onClick={(i) => this.handleClick(i)}
                />
                </div>
                <div className="game-info">
                    <div>{status}</div>
                    <div><ToggleButtonOnOff selected={this.state.historyOrder === 'ASC' ? true : false} onClick={() => this.changeHistoryOrder()} /></div>
                    <ol>{moves}</ol>
                </div>
            </div>
        );
    }
  }