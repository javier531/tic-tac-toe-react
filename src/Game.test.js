import React from 'react';
import {render, cleanup, act, fireEvent} from '@testing-library/react';
import {Game} from './game';
import {Square, Board} from './game-components';
import { calculateWinner } from './game-functions';

let container = null;

afterEach(cleanup);
beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});



it('check a snapshot', () => {
    const { asFragment } = render(<Game />);
    
    expect(asFragment(<Game />)).toMatchSnapshot();
});

it('check Square content should equal to X', () => {
    const square = render(<Square value='X' />); 
    expect(square.container.textContent).toBe('X');
});

it('check Board content', () => {
    let squareArray = Array(9).fill(null);
    squareArray[1] = squareArray[3] ='X';

    const board = render(<Board squares={squareArray} />); 
    const squares = board.getAllByText('X');
    expect(squares.length).toBe(2);
});

it('check Board click event', () => {
    const board = render(<Game />); 
    const squares = board.container.querySelectorAll('button.square');
    const square0 = squares.item(0);
    const square1 = squares.item(1);

    fireEvent.click(square0);
    fireEvent.click(square1);

    //Check first & second click
    expect(square0.textContent).toBe('X');
    expect(square1.textContent).toBe('O');
});

it('check toggle click event', () => {
    const board = render(<Game />); 
    const toggle = board.container.querySelector('button.toggle-button');

    expect(toggle.textContent).toBe('Orden ASC');
    fireEvent.click(toggle);

    //Check after click
    expect(toggle.textContent).toBe('Orden DESC');
});

it('check calculate winner', () => {
    let squares = null;
    const winners = [
        [0, 1, 2],
        [3, 4, 5],
        [6, 7, 8],
        [0, 3, 6],
        [1, 4, 7],
        [2, 5, 8],
        [0, 4, 8],
        [2, 4, 6],
      ];

      for(const winner of winners) {
        squares = Array(9).fill(null);
        squares[winner[0]] = 'X';
        squares[winner[1]] = 'X';
        squares[winner[2]] = 'X';
        expect(calculateWinner(squares)).toBe('X');
      }

      squares = Array(9).fill(null);
      squares[0] = 'X';
      squares[1] = 'X';
      squares[2] = 'O';
      expect(calculateWinner(squares)).toBe(null);

});