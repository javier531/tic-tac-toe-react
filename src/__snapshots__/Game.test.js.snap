// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check a snapshot 1`] = `
<DocumentFragment>
  <div
    class="game"
  >
    <div
      class="game-board"
    >
      <div>
        <div
          class="board-row"
        >
          <button
            class="sc-bdfBQB jOuJJT square"
          />
          <button
            class="sc-bdfBQB jOuJJT square"
          />
          <button
            class="sc-bdfBQB jOuJJT square"
          />
        </div>
        <div
          class="board-row"
        >
          <button
            class="sc-bdfBQB jOuJJT square"
          />
          <button
            class="sc-bdfBQB jOuJJT square"
          />
          <button
            class="sc-bdfBQB jOuJJT square"
          />
        </div>
        <div
          class="board-row"
        >
          <button
            class="sc-bdfBQB jOuJJT square"
          />
          <button
            class="sc-bdfBQB jOuJJT square"
          />
          <button
            class="sc-bdfBQB jOuJJT square"
          />
        </div>
      </div>
    </div>
    <div
      class="game-info"
    >
      <div>
        Pr√≥ximo: X
      </div>
      <div>
        <button
          class="toggle-button"
        >
          Orden ASC
        </button>
      </div>
      <ol>
        <li>
          <button
            class="last-move"
          >
            Reiniciar
          </button>
        </li>
      </ol>
    </div>
  </div>
</DocumentFragment>
`;
